File: chapter-02/exercise-01/indirection_expressions.cpp
#include <cassert>
#include <iostream>

int main() {
    int x = 10;
    int y = 20;
    long yield = 30;
    double miles = 40.5;
    int m = 50;
    int date = 2026;
    double distance = 60.25;
    char tab = '\t';
    int hours = 8;

    int* xAddr = &x;
    int* yAddr = &y;
    long* ptYld = &yield;
    double* ptMiles = &miles;
    int* mptr = &m;
    int* pdate = &date;
    double* distPtr = &distance;
    char* tabPt = &tab;
    int* hoursPt = &hours;

    assert(*xAddr == x);
    assert(*yAddr == y);
    assert(*ptYld == yield);
    assert(*ptMiles == miles);
    assert(*mptr == m);
    assert(*pdate == date);
    assert(*distPtr == distance);
    assert(*tabPt == tab);
    assert(*hoursPt == hours);

    std::cout << "a. *xAddr   = " << *xAddr << '\n';
    std::cout << "b. *yAddr   = " << *yAddr << '\n';
    std::cout << "c. *ptYld   = " << *ptYld << '\n';
    std::cout << "d. *ptMiles = " << *ptMiles << '\n';
    std::cout << "e. *mptr    = " << *mptr << '\n';
    std::cout << "f. *pdate   = " << *pdate << '\n';
    std::cout << "g. *distPtr = " << *distPtr << '\n';
    std::cout << "h. *tabPt   = tab character\n";
    std::cout << "i. *hoursPt = " << *hoursPt << '\n';
}

File: chapter-02/exercise-01/indirection_expressions.py
"""Demonstrate the Python equivalent of reading values through references.

The written answers use C++ dereference syntax such as *xAddr. Python does not
have that operator for ordinary variables; names and container entries already
hold references to objects.
"""

values = {
    "xAddr": 10,
    "yAddr": 20,
    "ptYld": 30,
    "ptMiles": 40.5,
    "mptr": 50,
    "pdate": 2026,
    "distPtr": 60.25,
    "tabPt": "\t",
    "hoursPt": 8,
}


def dereference(reference_name: str):
    """Return the object associated with a simulated reference name."""
    return values[reference_name]


def main() -> None:
    expected = [10, 20, 30, 40.5, 50, 2026, 60.25, "\t", 8]
    actual = [dereference(name) for name in values]
    assert actual == expected

    for name, value in values.items():
        display = "tab character" if value == "\t" else value
        print(f"{name:<8} -> {display}")


if __name__ == "__main__":
    main()
